exp/runtime/internal/controllers/extensionconfig_controller.go

```diff
@@ -43,23 +50,116 @@ func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, opt
+	if !extensionConfig.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, extensionConfig)
+	}
+
+	// discoverExtensionConfig will return a discovered ExtensionConfig with the appropriate conditions.
+	discoveredExtensionConfig, err := discoverExtensionConfig(ctx, r.RuntimeClient, extensionConfig)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	// Always patch the ExtensionConfig as it may contain updates in conditions.
+	if err = patchExtensionConfig(ctx, r.Client, extensionConfig, discoveredExtensionConfig); err != nil {
+		errs = append(errs, err)
+	}
+
+	if len(errs) != 0 {
+		return ctrl.Result{}, kerrors.NewAggregate(errs)
+	}
+
+	// Register the extensionConfig if it was found and patched without error.
+	if err = r.RuntimeClient.Register(discoveredExtensionConfig); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to register ExtensionConfig %s/%s", extensionConfig.Namespace, extensionConfig.Name)
+	}
 	return ctrl.Result{}, nil
 }
+
+func patchExtensionConfig(ctx context.Context, client client.Client, original, modified *runtimev1.ExtensionConfig, options ...patch.Option) error {
+	patchHelper, err := patch.NewHelper(original, client)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create PatchHelper for ExtensionConfig %s/%s", modified.Namespace, modified.Name)
+	}
+
+	options = append(options, patch.WithOwnedConditions{Conditions: []clusterv1.ConditionType{
+		runtimev1.RuntimeExtensionDiscoveredCondition,
+	}})
+	err = patchHelper.Patch(ctx, modified, options...)
+	if err != nil {
+		return errors.Wrapf(err, "failed to patch ExtensionConfig %s/%s", modified.Namespace, modified.Name)
+	}
+	return nil
+}
+
+// reconcileDelete will remove the ExtensionConfig from the registry on deletion of the object. Note this is a best
+// effort deletion that may not catch all cases.
+func (r *Reconciler) reconcileDelete(_ context.Context, extensionConfig *runtimev1.ExtensionConfig) (ctrl.Result, error) {
+	if err := r.RuntimeClient.Unregister(extensionConfig); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to unregister ExtensionConfig %s/%s", extensionConfig.Namespace, extensionConfig.Name)
+	}
+	return ctrl.Result{}, nil
+}
+
+// discoverExtensionConfig attempts to discover the Handlers for an ExtensionConfig.
+// If discovery succeeds it returns the ExtensionConfig with Handlers updated in Status and an updated Condition.
+// If discovery fails it returns the ExtensionConfig with no update to Handlers and a Failed Condition.
+func discoverExtensionConfig(ctx context.Context, runtimeClient runtimeclient.Client, extensionConfig *runtimev1.ExtensionConfig) (*runtimev1.ExtensionConfig, error) {
+	discoveredExtension, err := runtimeClient.Discover(ctx, extensionConfig)
+	if err != nil {
+		conditions.MarkFalse(extensionConfig, runtimev1.RuntimeExtensionDiscoveredCondition, runtimev1.DiscoveryFailedReason, clusterv1.ConditionSeverityError, "error in discovery: %v", err)
```

Comment:
I think this won't work. We are now modifying the input and output parameter again.

This means that the original we pass into `patchExtensionConfig` in l.113 will be modified here as well. 

(something like this was the reason we had the DeepCopy IIRC. I think we need either the DeepCopy here or one layer above)

Reply:
I think you're right. I'm going to change the input parameter here to be a non-pointer and add a deepcopy for the return.

Reply:
We discussed it a bit, (small) fix will follow later